### ISSUE TITLE
Supporting RUN instructions?

### ISSUE THREAD

**Issue by azymnis** [bazelbuild/rules_docker#90]
Skimming through the code, it does not seem that `docker_build` rules support RUN instructions: https://docs.docker.com/engine/reference/builder/#run

Is there a plan to add these?

Specifically I'd like to run `pip install -r requirements.txt`. For now my workaround is to build an image with my requirements pre-installed but that of course is very brittle

**Comment by mattmoor** [comment]
@duggelz @dlorenc @steren FYI

tl;dr `docker_build` avoids `RUN` by design because it is a huge can of non-determinism worms, but you should not need it for `pip install` (given more complete Bazel support).

Today Bazel has `py_binary`, `py_library`, and `py_test` rules that go in your `BUILD` files, but *should* (and does NOT) have a `pip_install` rule that goes in your `WORKSPACE` file.  

`WORKSPACE` is intended to capture your external dependencies and pull them down so that `BUILD` does not need anything but its declared dependencies (including a network connection!).  Given that Bazel is a universal build/test/packaging system, I like to think of `WORKSPACE` as its universal package manager client, where rules like `pip_install` are what should be talking to external package services and making those packages available to `BUILD`.

`git_repository` is easily one of the the most prominent "repository rules", but deals in source.  We define `docker_pull` here for downloading a base image.  Another good example in this space is `maven_jar` for downloading Java dependencies from Maven Central (or another Maven package mirror).

You can see some parallel discussion to this on pip [here](https://github.com/GoogleCloudPlatform/distroless/issues/69), I am going to resolve this and let that be the canonical issue tracking this.


It is a terrible fallback, but one option to express (a subset of) pip packages using Bazel's existing rules is [this](https://github.com/google/containerregistry/blob/master/WORKSPACE#L34).


I hope this answers your question and that we can get a repository rule for pip out sooner than later.  Thanks for reaching out.

**Comment by azymnis** [comment]
@mattmoor thanks for the explanation. I get it that RUN commands are a huge pain regarding non -determinism. One more (slightly) related question that I had was what to do regarding pulling debian packages that are not in the standard debian repos. In my case i need to pull a few packages for ROS (robot operating system). I can do this in a `Dockerfile` like this:

```
RUN sh -c 'echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -sc) main" > /etc/apt/sources.list.d/ros-latest.list'
RUN apt-key adv --keyserver hkp://ha.pool.sks-keyservers.net:80 --recv-key 421C365BD9FF1F717815A3895523BAEEB01FA116
RUN apt-get update
RUN apt-get install ros-jade-rospy-message-converter
```

How would I do something like that in a docker bazel rule?

**Comment by mattmoor** [comment]
@r2d4 owns some `deb` rules in the `distroless` repo.

@dlorenc FYI

**Comment by jschenxiaoyu** [comment]
@mattmoor I have a problem when using `uwsgi`. I have to install it by `pip install uwsgi` and then start the server using `uwsgi xxx` command. So how can I start `uwsgi` service using `rules_docker`?

**Comment by mattmoor** [comment]
Can you open a separate issue?  I think the tl;dr is `entrypoint`, but there are a lot of things to get wrong before that.

**Comment by fxer** [comment]
@azymnis you ever get an answer to the proper method to `apt install` something into a bazel built container?

**Comment by azymnis** [comment]
@fxer no. What I end up doing is just creating base images with any Debian packages that I need and just rooting my docker bazel build rules from there. Does that make sense?

**Comment by nlopezgi** [comment]
@fxer  download_pkgs and install_pkgs might work for your use case -
https://github.com/GoogleContainerTools/base-images-docker/tree/master/package_managers

**Comment by cnsgsz** [comment]
With [bazel's support for pip_install](https://github.com/bazelbuild/rules_python#using-the-packaging-rules), is the recommended  to handle python dependencies with that rather than rules_docker (e.g. `cmd = ["pip", "install", "-r", "requirement.txt"]), files="requirement.txt"`)?
